### PR TITLE
Support datasets with mixed magnifications

### LINKED ISSUE #4018: Texture count is not used per layer in fragment shader
Instead, the maximum is used which works per se, but can cause unnecessary branching --> performance.

### LINKED ISSUE #6891: Support mixed magnifications in different layers
For example, mag 2-2-1 and mag 2-2-2 (different layers, though).

### PR DESCRIPTION
**Pull request by philippotto**
### URL of deployed dev instance (used for testing):
- https://mixedmags.webknossos.xyz

### Steps to test:
- download and setup dataset from here: https://www.notion.so/scalableminds/Test-Datasets-c0563be9c4a4499dae4e16d9b2497cfb?pvs=4#71e2b67286dd4cdf99cb1164ca97c0e0
- view dataset
- create annotation
- monkey-test

### Todo:
- [x] back-end should to default to the union of mags when creating a new segmentation layer (without fallback) IFF the union doesn't contain ambiguous mags for a mag level

### Issues:
- fixes #6891
- fixes #4018

------
(Please delete unneeded items, merge only when none are left open)
- [x] Updated [changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Updated [migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [ ] Updated [documentation](../blob/master/docs) if applicable
- [ ] Adapted [wk-libs python client](https://github.com/scalableminds/webknossos-libs/tree/master/webknossos/webknossos/client) if relevant API parts change
- [ ] Removed dev-only changes like prints and application.conf edits
- [ ] Considered [common edge cases](../blob/master/.github/common_edge_cases.md)
- [ ] Needs datastore update after deployment

### CODE REVIEW

**Comment by fm3** [comment]
Sounds good :) I think both versions for the auto selection are viable. Do you have a preference? Did you already make a choice for the front-end (modal) selection?

**Comment by philippotto** [comment]
> Do you have a preference? Did you already make a choice for the front-end (modal) selection?

The frontend currently picks the first possible mag for each mag level (ordered according to layer order). See this code here:

https://github.com/scalableminds/webknossos/blob/619d9681ba5f09a34c1a56ca2fcd75bd552f2659/frontend/javascripts/oxalis/model/accessors/dataset_accessor.ts#L92

Doing the same in the back-end is probably the right call.

I'm still a bit torn on this. On the one hand, mags will usually match and this heuristic for the mag restrictions is probably fine for now. On the other hand, I'm wondering if we should do it correctly now..

**Comment by fm3** [comment]
what do you think doing it correctly now would mean? :grimacing:

**Comment by philippotto** [comment]
@daniel-wer I think the front-end is already ready for review. The back-end needs a small tweak, but this shouldn't impact the front-end.

**Review by daniel-wer — COMMENTED** [review]
Well done, nice to see that webknossos as a whole was fairly well prepared for this, not many large changes were needed.
Out of curiosity, you removed quite a bit of dead code - Did you use a specific tool to identify it?

I started with testing but will continue tomorrow. For now only one note:
- The info tab tooltip showing rendered mags per layer should use the display name of the volume layer, see screenshot
image

**Comment by daniel-wer** [comment]
I did some more testing and everything worked very well :tada:

File: frontend/javascripts/oxalis/geometries/materials/plane_material_factory.ts [hunk 1]
```diff
@@ -394,6 +403,46 @@ class PlaneMaterialFactory {
...
+          // To align the vertices with the buckets, the current magnification is
+          // needed. Since the current mag can differ from layer to layer, the shader
+          // needs to know which mag is safe to use.
+          // For this purpose, we define the representativeMagForVertexAlignment which is
+          // an virtual mag (meaning, there's not necessarily a layer with that exact
+          // mag). It is derived from the layers that are not transformed by considering
+          // the minimum for each axis. That way, the vertices are aligned using the
+          // lowest common multiple.
+          // For example, one layer might render mag 4-4-1, whereas another layer renders
+          // 4-2-1. The representative mag would be 1-2-1.
```
[the remark this answers is not in the document]
Comment (philippotto) [reply frontend/javascripts/oxalis/geometries/materials/plane_material_factory.ts:420]: You're right, I integrated your suggestion :+1:

File: frontend/javascripts/dashboard/advanced_dataset/create_explorative_modal.tsx [hunk 2]
```diff
@@ -192,24 +193,19 @@ function CreateExplorativeModal({ datasetId, onClose }: Props) {
...
-      );
-      highestResolutionIndex = datasetFallbackLayerResolutionInfo.getHighestResolutionIndex();
-      lowestResolutionIndex = datasetFallbackLayerResolutionInfo.getClosestExistingIndex(0);
-    }
+    const resolutionInfo =
+      selectedSegmentationLayer == null
+        ? getSomeResolutionInfoForDataset(dataset)
+        : getResolutionInfo(selectedSegmentationLayer.resolutions);
+    const highestResolutionIndex = resolutionInfo.getHighestResolutionIndex();
+    const lowestResolutionIndex = resolutionInfo.getClosestExistingIndex(0);
```
[the remark this answers is not in the document]
Comment (philippotto) [reply frontend/javascripts/dashboard/advanced_dataset/create_explorative_modal.tsx:201]: Good point, there shouldn't be any difference. I replaced all occurrences of `getClosestExistingIndex(0);` with `getLowestResolutionIndex()`.

**Comment by philippotto** [comment]
@daniel-wer thanks for your review! could you have another look at my most recent commits? :)

**Pull request merged**